src/main/java/com/linkedin/kafka/clients/largemessage/Version0Deseriaizer.java

```diff
@@ -0,0 +1,95 @@
+package com.linkedin.kafka.clients.largemessage;
+
+import com.linkedin.kafka.clients.headers.HeaderDeserializer;
+import com.linkedin.kafka.clients.headers.HeaderUtils;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Attempts to deserialize the current version and then the previous version that did not use headers for large messages.
+ * <pre>
+ * The large message segment serializer.
+ * The format of the serialized segment is:
+ * 1 byte   - version
+ * 4 bytes  - checksum to determine if bytes is large message segment or not.
+ * 16 bytes - messageId
+ * 4 bytes  - sequence number
+ * 4 bytes  - number of segments
+ * 4 bytes  - message size in bytes
+ * X bytes  - payload
+ * </pre>
+ */
+public class Version0Deseriaizer implements HeaderDeserializer {
+
+  private final static Logger LOG = LoggerFactory.getLogger(Version0Deseriaizer.class);
+  static final int SEGMENT_INFO_OVERHEAD = 16 + Integer.BYTES + Integer.BYTES + Integer.BYTES;
+  static final byte LARGE_MESSAGE_ONLY_VERSION = 0;
+  static final int CHECKSUM_LENGTH = Integer.BYTES;
+
+  private final HeaderDeserializer _tryFirst;
+
+  public Version0Deseriaizer(HeaderDeserializer tryFirst) {
+    if (tryFirst == null) {
+      throw new IllegalArgumentException("This deserializer only supports legacy serilized values.  No new values of"
+          + " this type are supported.");
+    }
+    _tryFirst = tryFirst;
+  }
+
+  @Override
+  public DeserializeResult deserializeHeader(ByteBuffer src) {
+    DeserializeResult deserializeResult = _tryFirst.deserializeHeader(src);
+    if (src == null && deserializeResult.value() == null) {
+      return deserializeResult;  //null value
+    }
+    if (deserializeResult.headers() != null || deserializeResult.value() != src) {
```

Comment:
Yeah this is kind of hacky.  I've introduced a success flag in the returned result.